app/models/invoice.rb

```diff
@@ -248,8 +249,50 @@ def void_invoice!
   def ensure_number
     return if number.present?
 
-    formatted_sequential_id = format('%03d', sequential_id)
+    if organization.document_numbering.to_s == 'per_customer'
+      customer_slug = "#{organization.document_number_prefix}-#{format('%03d', customer.sequential_id)}"
+      formatted_sequential_id = format('%03d', sequential_id)
 
-    self.number = "#{customer.slug}-#{formatted_sequential_id}"
+      self.number = "#{customer_slug}-#{formatted_sequential_id}"
+    else
+      org_formatted_sequential_id = format('%03d', organization_sequential_id)
+
+      self.number =
+        "#{organization.document_number_prefix}-#{Time.now.utc.strftime('%Y%m')}-#{org_formatted_sequential_id}"
+    end
+  end
+
+  def ensure_organization_sequential_id
+    return if organization_sequential_id.present? && organization_sequential_id.positive?
+
+    self.organization_sequential_id = generate_organization_sequential_id
+  end
+
+  def generate_organization_sequential_id
+    organization_sequence_scope = organization.invoices.where(created_at: Time.now.utc.all_month)
+
+    result = Invoice.with_advisory_lock(
```

Comment:
could this code be moved into the sequenced concern?

Reply:
As described offline, if you agree, I would rather keep this code portion in invoice model since sequenced concern is also included in few other models (credit note & customer) where this logic is not needed

Reply:
Okay, maybe it will be good to improve this part later (but not so late)